pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java

```diff
@@ -333,6 +335,7 @@ public void sendAsync(Message<T> message, SendCallback callback) {
                     // batch size and/or max message size
                     if (batchMessageContainer.hasSpaceInBatch(msg)) {
                         batchMessageContainer.add(msg, callback);
+                        lastSendFuture = callback.getFuture();
```

Comment:
Can't we just assign lastSendFuture once, before calling isBatchMessagingEnabled()? Once we get to that point, the message is going to best sent in any case.

Reply:
fixed